server/src/test/scala/com/ossum/riddl/lsp/server/RiddlLSPTextDocumentSpec.scala

```diff
@@ -0,0 +1,197 @@
+      textDocumentItem
+    )
+
+    service.didOpen(openNotification)
+  }
+
+  trait OpenEmptyFileSpec extends EmptyInitializeSpec {
+    val openNotification: DidOpenTextDocumentParams =
+      new DidOpenTextDocumentParams()
+    openNotification.setTextDocument(
+      textDocumentItem
+    )
+
+    service.didOpen(openNotification)
+  }
+
+  trait ChangeOneErrorFileSpec extends OpenOneErrorFileSpec {
+    val changeNotification: DidChangeTextDocumentParams =
+      new DidChangeTextDocumentParams()
+
+    val versionedDocIdentifier = new VersionedTextDocumentIdentifier()
+    versionedDocIdentifier.setUri(docURI)
+    changeNotification.setTextDocument(versionedDocIdentifier)
+
+    val changes = new TextDocumentContentChangeEvent()
+    changes.setText("the")
+    val changeRange = new lsp4j.Range()
+
+    val changeStart = new Position()
+    changeStart.setLine(11)
+    changeStart.setCharacter(21)
+    changeRange.setStart(changeStart)
+
+    val changeEnd = new Position()
+    changeEnd.setLine(11)
+    changeEnd.setCharacter(24)
+    changeRange.setEnd(changeEnd)
+
+    changes.setRange(changeRange)
+    changeNotification.setContentChanges(List(changes).asJava)
+
+    service.didChange(changeNotification)
+  }
+
+  trait ChangeEmptyFileSpec extends OpenEmptyFileSpec with OneErrorFileSpec {
+    val changeNotification: DidChangeTextDocumentParams =
+      new DidChangeTextDocumentParams()
+
+    val versionedDocIdentifier = new VersionedTextDocumentIdentifier()
+    versionedDocIdentifier.setUri(docURI)
+    changeNotification.setTextDocument(versionedDocIdentifier)
+
+    val changes = new TextDocumentContentChangeEvent()
+    changes.setText("domain New {}")
+    val changeRange = new lsp4j.Range()
+
+    val changeStart = new Position()
+    changeStart.setLine(1)
+    changeStart.setCharacter(1)
+    changeRange.setStart(changeStart)
+
+    val changeEnd = new Position()
+    changeEnd.setLine(1)
+    changeEnd.setCharacter(1)
+    changeRange.setEnd(changeEnd)
+
+    changes.setRange(changeRange)
+    changeNotification.setContentChanges(List(changes).asJava)
+
+    service.didChange(changeNotification)
+  }
+
+  "RiddlLSPTextDocumentService" must {
+    "successfully open everythingOneError.riddl & get completion" in new OpenOneErrorFileSpec {
+      val params = new CompletionParams()
+      params.setTextDocument(textDocumentIdentifier)
+      val position = new Position()
+      position.setLine(5)
+      position.setCharacter(7)
+      params.setPosition(position)
+
+      val resultF: CompletableFuture[
+        messages.Either[util.List[CompletionItem], CompletionList]
+      ] = service.completion(params)
+
+      whenReady(resultF.asScala) { eitherList =>
+        eitherList.isRight mustBe true
+        eitherList.getRight.getItems.asScala.length mustEqual 1
+        eitherList.getRight.getItems.asScala.head.getTextEditText mustEqual
+          """Expected one of (end-of-input | whitespace after keyword)"""
```

Comment:
Good.  Is there anything else you can assert from the error messages? 
Here is the message class:  `Message(loc: At, message: String, kind: KindOfMessage = Error, context: String = "")`
[from here](https://github.com/ossuminc/riddl/blob/main/language/src/main/scala/com/ossuminc/riddl/language/Messages.scala#L110). You could check the context and the location The At class is in the parsing subdirectory

Reply:
`Location` is used to request the `Message` in the first place here. 

In fact, `Message` is not part of the response, as we get back (here) `Either[List[CompletionItem], CompletionList]`, which I am comparing with the `Message.message` to make my assertion. Note that I have no clue why LSP4J uses an `Either` here, it seems highly redundant, probably an API backward compatibility thing but oh well. I am only using the `Right` side of the `Either` in my responses, ie a `List[CompletionItem]` as a Left should be inferred as an error result by the client.

I can search for a function that reports back or takes in `context` as a field, but I think that should wait for a subsequent PR.

Reply:
Okay. The text you're comparing from comes from the RIDDL parser so not sure how you got LSP to run the RIDDL parser. *shrug*.

Reply:
The `CompletionItem` is seen below printed from the end of the test. It contains many fields that can be tested, but I am an not sure what is important here or even possible to fill out.

```
CompletionItem [
  label = null
  labelDetails = null
  kind = null
  tags = null
  detail = null
  documentation = null
  deprecated = null
  preselect = null
  sortText = null
  filterText = null
  insertText = null
  insertTextFormat = null
  insertTextMode = null
  textEdit = null
  textEditText = "Expected one of (end-of-input | whitespace after keyword)"
  additionalTextEdits = null
  commitCharacters = null
  command = null
  data = null
]
```

Reply:
@reid-spencer I am using TopLevelParser from String or URL source, currently always the latter (`fromUrl` is never `false`:

private def updateParsedDoc(fromURL: Boolean = true): Unit = {
    if fromURL then docAST = docURI.map(parseDocFromSource)
    else docAST = riddlDoc.map(parseDocFromString)

    updateDocLines()
}
  
def parseDocFromSource(docURI: String): Either[Messages, AST.Root] = {
  val riddlRootDoc = io.Source.fromURL(docURI)
  new TopLevelParser(SourceParserInput(riddlRootDoc, docURI)).parseRoot()
}

def parseDocFromString(docString: String): Either[Messages, AST.Root] =
  new TopLevelParser(StringParserInput(docString)).parseRoot()

Reply:
> The `CompletionItem` is seen below printed from the end of the test. It contains many fields that can be tested, but I am an not sure what is important here or even possible to fill out. I am noticing now there is a `CompletionItem.kind` - but this in LSP4J is more like a kind of riddl definition than a kind of error. `Message.kind` could be added as a tag though, that's about the best place I see to put it. As for `Message.context`, I am not sure how this will translate either
> 
> ```
> CompletionItem [
>   label = null
>   labelDetails = null
>   kind = null
>   tags = null
>   detail = null
>   documentation = null
>   deprecated = null
>   preselect = null
>   sortText = null
>   filterText = null
>   insertText = null
>   insertTextFormat = null
>   insertTextMode = null
>   textEdit = null
>   textEditText = "Expected one of (end-of-input | whitespace after keyword)"
>   additionalTextEdits = null
>   commitCharacters = null
>   command = null
>   data = null
> ]
> ```

I would suggest that you're providing it the parser's text output rather than the input.  GIven that the textEditText is the only non-null field, it looks pretty useless.

Reply:
> @reid-spencer I am using TopLevelParser from String or URL source, currently always the latter (`fromUrl` is never `false`):
> 
> ```
> private def updateParsedDoc(fromURL: Boolean = true): Unit = {
>     if fromURL then docAST = docURI.map(parseDocFromSource)
>     else docAST = riddlDoc.map(parseDocFromString)
> 
>     updateDocLines()
> }
>   
> def parseDocFromSource(docURI: String): Either[Messages, AST.Root] = {
>   val riddlRootDoc = io.Source.fromURL(docURI)
>   new TopLevelParser(SourceParserInput(riddlRootDoc, docURI)).parseRoot()
> }

The SourceParser might be problematic. It isn't heavily tested and those classes aren't supposed to be used directly. In fact, I should make them private.  Have you tried:
```
val input = RiddlParserInput(  java.net.URI.create(docURI).toURL )
new TopLevelParser(input).parseRoot
```
?
> 
> def parseDocFromString(docString: String): Either[Messages, AST.Root] =
>   new TopLevelParser(StringParserInput(docString)).parseRoot()
> ```

Reply:
The exact same result was obtained using `RiddlParserInput`. If you think this is safer, I will keep the change.

I think there is still a misunderstanding around `CompletionItem`: this is the `completions` function response, and so must be used as the form of the response from the Server. In other words, we must translate `Message`s to this form. 

This is a function for responding with the `completion` suggestion when it is requested for a word at a given line & character

Reply:
I just asked Bing, it suggested using `details` or `documentation` instead of `textEditText`. Right now, I don't see any way to suggest how to fix the error to the user, only how to report the error.

Reply:
Yes, that will be safer. I just made the RiddlParserInput subclasses inaccessible (if you build from main)